internal/rest/rest.go

```diff
@@ -23,37 +24,23 @@ import (
 )
 
 func handleAPIRequest(action rest.EndpointAction, state *state.State, w http.ResponseWriter, r *http.Request) response.Response {
-	trusted := r.Context().Value(request.CtxAccess)
-	if trusted == nil {
-		return response.Forbidden(nil)
-	}
-
-	trustedReq, ok := trusted.(internalAccess.TrustedRequest)
-	if !ok {
-		return response.Forbidden(nil)
-	}
-
-	// If the request is not trusted, only allow it if AllowUntrusted is set.
-	if !trustedReq.Trusted && !action.AllowUntrusted {
-		return response.Forbidden(nil)
-	}
-
 	if action.Handler == nil {
 		return response.NotImplemented(nil)
 	}
 
-	if action.AccessHandler != nil {
-		// Defer access control to custom handler.
-		accessResp := action.AccessHandler(state, r)
-		if accessResp != response.EmptySyncResponse {
-			return accessResp
+	// If allow untrusted is not set, the request must be authenticated via core authentication (e.g. certificate in truststore).
+	if !action.AllowUntrusted {
```

Comment:
If the downstream app opt to implement only oidc authentication, do we then set every endpoint to AllowUntrusted=true and then pass a custom AccessHandler?

Unrelated to this line, is it okay if we pass w http.ResponseWriter to the handler functions? Looking at implementing a oidc/login endpoint in lxd-site-manager where I need to set some cookies on the response.

Reply:
> If the downstream app opt to implement only oidc authentication, do we then set every endpoint to AllowUntrusted=true and then pass a custom AccessHandler?

Yep that's how it would work currently. The discussion above was around this. The `AllowUntrusted` field now means "allow requests that are not TLS or are TLS but where the public key of he requestor is not in the microcluster truststore".

For this, you can use a `response.ManualResponse`. I'll show you an example when we have a call on Monday :)

Reply:
Amazing, thanks for the clarification! I will have a look at `response.ManualResponse` so long :+1: